packages/generator/src/generator.js

```diff
@@ -217,6 +218,77 @@ export default class Generator {
       consola.warn('HTML minification failed for SPA fallback')
     }
 
+    const { csp } = this.options.render
```

Comment:
I think there is CSP string in response headers of `renderRoute` returned value, we may use it to generate `meta`.

Reply:
I've checked. `this.nuxt.server.renderRoute(...)` returns an object with only two items:
* `html` (string): the rendered HTML
* `preloadFiles` Array of Objects, each representing one of the key JS files

No CSP headers are returned.

Afaics `renderRoute()` calls different renderer for SSR and SPA. While the `SSRRenderer`'s `render()` method [really returns](https://github.com/nuxt/nuxt.js/blob/dev/packages/vue-renderer/src/renderers/ssr.js#L285) `cspScriptSrcHashes`, the `SPARenderer`'s `render()` [only returns](https://github.com/nuxt/nuxt.js/blob/dev/packages/vue-renderer/src/renderers/spa.js#L174) the html and the files.

Reply:
Sorry, csp in response only works in ssr request.

I think a proper way is supporting csp meta tag in spa renderer and also extract csp hash generation and getCspString code.

Reply:
That was my first approach, but it is impossible, since the rendered HTML is minified after being returned [here](https://github.com/nuxt/nuxt.js/blob/dev/packages/generator/src/generator.js#L215) so any hashes calculated during generation will become invalid after minification. That is why I've put the CSP generation code into the generator, so it can act on the minified JS and CSS tags.